Firestore/Swift/Source/ReferenceableObject/ReferenceableObjectManager.swift

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import FirebaseFirestore
+
+import CryptoKit
+import OSLog
+
+/// Used to fetch and save ReferenceableObjects.
+///
+///  To prevent refetch of the same referenced object immediately, the manager
+///  also momentarily caches the referenced object. This interval is configurable.
+///
+///  To prevent writes of unmodified referenced objects, the manager compares checksums for the
+///  object being written.
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+public class ReferenceableObjectManager {
+  public static var instance = ReferenceableObjectManager()
+
+  static var settings = ReferenceableObjectManagerSettings()
+
+  private var db = Firestore.firestore()
+
+  private var objectCache = ReferenceableObjectCache()
+
+  public func save<T: ReferenceableObject>(object: T) async throws {
+    do {
+      let encoder = Firestore.Encoder()
+      let json = try encoder.encode(object)
+
+      if let docId = object.id {
+        guard let currentDigest = computeHash(obj: json),
+              await needsSave(object: object, currentDigest: currentDigest) else {
+          OSLog.storeByReference.debug("Object doesn't need to be saved")
+          return
+        }
+
+        try await db.collection(T.parentCollection()).document(docId).setData(json)
+        await objectCache.add(object: object, digest: currentDigest)
+      } else {
+        try await db.collection(T.parentCollection()).document().setData(json)
+      }
+
+      print("Save complete")
```

Comment:
Might want to use the logger.

Reply:
Done. I also moved the refactored the Logger class so it could be reused across FirestoreSwift if need be.